### PR TITLE
Pass `RouteManager` as proxy for router.addRoutes

### LINKED ISSUE #3479: Pass `RouteManager` additionally to router to modules
## What is the motivation for adding / enhancing this feature?

@filrak currently, because of Multistore and optimizations made in #3443 we don't want users to call `router.addRoutes` directly; we should then pass `RouteManager` instead of plain `vue-router` into module constructors - especially that we have more/less the same interface so it's not a breaking change even

related to #3443 
related to #3454

### PR DESCRIPTION
**Pull request by gibkigonzo**
### Related issues

closes #3479

### Short description and why it's useful

Add proxy for `vue-router` and pass it for other files instead of `RouterManager.addRoutes`.

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [x] I've updated [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by pkarw** [comment]
@filrak OK cool; @gibkigonzo please add a note to Upgrade Notes (doc/upgrade-notes/README.md) about the solution for those who must support IE11 (with the https://github.com/GoogleChrome/proxy-polyfill)

**Comment by pkarw** [comment]
@kkdg this might be imoportant note (about the Proxies) regarding the upgrade notes

**Comment by pkarw** [comment]
OK guys, I've checked the situations across the known projects going on VS right now - we can't apply this change; we must have transpilable proxy @gibkigonzo can you please apply the changes?

**Comment by gibkigonzo** [comment]
I've added Proxy polyfill if native doesn't exist. Also I have to add new type `VSFRouter` to override VueRouter. I didn't found better solution to override `addRoutes` then using overload.

**Comment by filrak** [comment]
I'm worried changing the interface name may break some 3rd parties relying on this. How about just extending the VueRouter declaration (check `shims.d.ts' and https://www.typescriptlang.org/docs/handbook/declaration-merging.html )

File: core/app.ts [hunk 1]
```diff
@@ -49,16 +51,36 @@ function createRouter (): VueRouter {
   })
 }
 
+function createRouterProxy (router: VueRouter): VSFRouter {
```
Comment (andrzejewsky) [inline core/app.ts]: Can you maybe use arrow function to keep consistence with the rest of file?

Comment (andrzejewsky) [reply core/app.ts]: And also you can move this function to the `core/helpers` and import here

File: src/modules/amp-renderer/index.ts [hunk 2]
```diff
@@ -14,5 +13,5 @@ const ampRendererStore = {
 export const AmpRendererModule: StorefrontModule = function (app, store, router, moduleConfig, appConfig) {
   store.registerModule('amp-renderer', ampRendererStore)
   setupMultistoreRoutes(config, router, moduleRoutes, 10)
-  // RouterManager.addRoutes(moduleRoutes, router, true)
+  // router.addRoutes(moduleRoutes, router, true)
```
Comment (andrzejewsky) [inline src/modules/amp-renderer/index.ts]: Do you need this comment here?

File: src/themes/default-amp/index.js [hunk 3]
```diff
@@ -9,5 +8,5 @@ export default function (app, router, store) {
   // To do so please exclude the desired storeView from the config.storeViews.mapStoreUrlsFor and map the urls by your own like:
   // { name: 'de-checkout', path: '/checkout', component: CheckoutCustomized },
   setupMultistoreRoutes(config, router, routes, 10)
-  // RouterManager.addRoutes(routes, router, true)
+  // router.addRoutes(routes, router, true)
```
Comment (andrzejewsky) [inline src/themes/default-amp/index.js]: and here?

**Comment by pkarw** [comment]
@gibkigonzo please apply the @andrzejewsky comments in order to merge this one

**Comment by gibkigonzo** [comment]
@filrak I've checked that `declaration merging` and it looks very nice :) I can add something like this:
```
declare module 'vue-router' {
  import VueRouter, { RouteConfig } from 'node_modules/vue-router'
  export * from 'node_modules/vue-router'
  export default class extends VueRouter {
    public addRoutes (routes: RouteConfig[], useRouteQueue?: boolean, priority?: number): void;
  }
}
```
and `VSFRouter` will be unnecessary :)

**Comment by pkarw** [comment]
@gibkigonzo please let us know when it'll be ready to merge

File: src/modules/amp-renderer/index.ts (hunk 2 above)
Comment (gibkigonzo) [reply src/modules/amp-renderer/index.ts]: nope, I've removed it

File: src/themes/default-amp/index.js (hunk 3 above)
Comment (gibkigonzo) [reply src/themes/default-amp/index.js]: same as above :)

File: core/app.ts (hunk 1 above)
Comment (gibkigonzo) [reply core/app.ts]: good point, I've change it

**Comment by gibkigonzo** [comment]
@pkarw I've applied suggestions and it can be merged

**Comment by andrzejewsky** [comment]
thanks @gibkigonzo I like this workaround to keep types compatibility, good job 👍

**Pull request merged**